### PR TITLE
Fix filter by branch with globbing characters *?[]

### LINKED ISSUE #8800: Filter by branches does not apply if contains globbing characters [?*
## Current behaviour

- Type `branch_name_prefix_*` into branches combobox text area, press Enter.

Actual result: empty revision grid

## Expected behaviour
Revision grid is filtered to show commits from any branch with name starting with `branch_name_prefix_`

## Screenshots

Before I input the filter, you can see there is a branch named `CSMS-9313_migrate_sender_fee`

image

Then I input a filter with glob character by `CSMS-9313_*`

image

## Related issues

While fixing this, maybe we can support filter by multiple branches as well

https://github.com/gitextensions/gitextensions/issues/491
https://github.com/gitextensions/gitextensions/issues/3546

## Diagnostics

The actual code is aware of globbing characters, the branches filter with globbing is converted to `--branches=...` parameter of `git log` command

https://github.com/gitextensions/gitextensions/blob/fb0aa6d06a2c78a5f25ea6a9e351d505e85be638/GitCommands/RevisionReader.cs#L194

However `refFilterOptions.HasFlag(RefFilterOptions.Branches)` is never `true`, so the option `--branches` is never added.

### PR DESCRIPTION
**Pull request by NikolayXHD**
Fixes #8800

## Test methodology

Manually

## Test environment(s)

Windows 10

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/RevisionReader.cs [hunk 1]
```diff
@@ -172,44 +172,46 @@ public sealed class RevisionReader : IDisposable
...
+            var builder = new GitArgumentBuilder("log");
+            builder.Add("-z");
+            builder.Add($"--pretty=format:\"{FullFormat}\"");
+            if (!string.IsNullOrWhiteSpace(branchFilter) && isBranchFilterSimple(branchFilter))
             {
-                "-z",
-                branchFilter,
-                $"--pretty=format:\"{FullFormat}\"",
+                builder.Add(branchFilter);
+            }
```
Comment (RussKie) [inline GitCommands/RevisionReader.cs:181]: The original syntax was terse, and thus I'm finding it a bit more readable.

What's wrong with the following?
```cs
            return new GitArgumentBuilder("log")
            {
                "-z",
                branchFilter,
                { !string.IsNullOrWhiteSpace(branchFilter) && isBranchFilterSimple(branchFilter), branchFilter }
            }
```

File: GitCommands/RevisionReader.cs [hunk 2]
```diff
@@ -172,44 +172,46 @@ public sealed class RevisionReader : IDisposable
...
+                { refFilterOptions.HasFlag(RefFilterOptions.NoMerges), "--no-merges" },
+                { refFilterOptions.HasFlag(RefFilterOptions.SimplifyByDecoration), "--simplify-by-decoration" }
+            }.ToString());
+            builder.Add(revisionFilter);
+            builder.Add("--");
+            builder.Add(pathFilter);
+            return builder;
         }
 
+        private static bool isBranchFilterSimple(string branchFilter) =>
```
Comment (RussKie) [inline GitCommands/RevisionReader.cs:212]: [naming]
```suggestion
        private static bool IsSimpleBranchFilter(string branchFilter) =>
```

File: GitCommands/RevisionReader.cs (hunk 1 above)
Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:181]: The sequence `"-z", branchFilter` is only valid if the branch filter is simple

```bash
# valid
git log -z full_branch_name
git log --branches=*_name

#invalid
git log -z *-name
git log -z "*-name"
```

Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:181]: thus I need the `if` which breaks initializer-like syntax:
```
if (is_simple_branch_filter)
  builder.Add(branchFilter);
```

Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:181]: Somehow I missed the possibility to
```
{boolean_condition, the_flag}
```
I was under impression, that the only possibility of conditional flag was ternary semantics
```
{boolean_condition, flag_if_true, flat_if_false}
```

File: GitCommands/RevisionReader.cs [hunk 3]
```diff
@@ -172,44 +172,46 @@ public sealed class RevisionReader : IDisposable
...
-            return new GitArgumentBuilder("log")
+            var builder = new GitArgumentBuilder("log");
+            builder.Add("-z");
+            builder.Add($"--pretty=format:\"{FullFormat}\"");
+            if (!string.IsNullOrWhiteSpace(branchFilter) && isBranchFilterSimple(branchFilter))
             {
-                "-z",
-                branchFilter,
-                $"--pretty=format:\"{FullFormat}\"",
+                builder.Add(branchFilter);
```
Comment (gerhardol) [inline GitCommands/RevisionReader.cs:180]: --branches accepts wildcards. Is it not sufficient to remove the check for special chars?

Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:180]: --branches option [has a weird behavior](https://www.git-scm.com/docs/git-log#Documentation/git-log.txt---branchesltpatterngt), when no wildcards are present in the branchFilter:

>  If pattern lacks ?, *, or [, /* at the end is implied

This is why we have to check if these characters are in branch filter, and where they are not, we use `git log branchFilter` instead

File: GitCommands/RevisionReader.cs (hunk 1 above)
Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:181]: fixed

File: GitCommands/RevisionReader.cs (hunk 2 above)
Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:212]: ok

File: GitCommands/RevisionReader.cs [hunk 4]
```diff
@@ -175,8 +175,11 @@ public sealed class RevisionReader : IDisposable
             return new GitArgumentBuilder("log")
             {
                 "-z",
-                branchFilter,
                 $"--pretty=format:\"{FullFormat}\"",
+                {
```
Comment (gerhardol) [inline GitCommands/RevisionReader.cs:179]: Any reason to switch order?

I would pprefer to move this option close to the other where branchFilter is handled (either here or to line 194)

**Review by gerhardol — COMMENTED** [review]
+1
You get "no results" with wildcards. Not really a problem.
Useful.

image

File: GitCommands/RevisionReader.cs (hunk 4 above)
Comment (NikolayXHD) [reply GitCommands/RevisionReader.cs:179]: > Any reason to switch order?

No reason, my mistake. Moved back to where it was.

**Pull request merged**